src/email/email.service.ts

```diff
@@ -0,0 +1,54 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { MailerService } from '@nestjs-modules/mailer';
+import { TokensService } from 'src/tokens/tokens.service';
+import { User } from 'src/users/entities/user.entity';
+import { MailSubject, MailTemplate } from 'src/utils/mail/mail-details';
+
+import { RegistrationLinks, TokenData, TokenOptions } from '../types';
+
+@Injectable()
+export class EmailService {
+  constructor(
+    private readonly mailerService: MailerService,
+    private readonly tokensService: TokensService,
+    private readonly configService: ConfigService,
+  ) {}
+
+  private async createUrls(users: User[], tokenType: TokenOptions): Promise<RegistrationLinks[]> {
+    const registrationLinks: RegistrationLinks[] = [];
+    const promises = users.map(async user => {
+      const tokenData: TokenData = await this.tokensService.createToken(user, tokenType);
+      const { token } = tokenData;
+      const url = `http://localhost:3000/api/auth/register/${user.id}/${token}`;
+      registrationLinks.push({
+        email: user.email,
+        url,
+      });
+    });
+    await Promise.all(promises);
+    Logger.log('users', users, 'tokenType', tokenType);
+    return registrationLinks;
```

Comment:
Tutaj podszedłbym do tematu trochę inaczej. 
stworzyłbym samodzielną metodę w `TokensService` do tworzenia/generowania linku (w zależności od rodzaju tokena - póki co mamy aktywacyjny i resetujący hasło) - Możesz też w module `TokensModule` stworzyć oddzielny serwis np. `LinkService`, który jest odpowiedzialny tylko za obsługę takich linków.

Potem tworzymy ładną metodę generującą tablicę obiektów {email, link} (tutaj super do tego podszedłeś:) )i też bym ją wtedy zostawił w `LinkService` (`LinksService` - nigdy nie wiem czy to powinno być w liczbie pojedynczej, czy mnogiej).